File: packages/markets/src/tradewinds/markets/catalog/kalshi_nhigh.py
"""Kalshi NHIGH contract spec (daily HIGH temperature settlement).

NHIGH markets resolve against the NWS CLI ``max_temp_f`` value for a
specific station on a specific date. ``resolve(contract_id, date)`` is
the deterministic mapping from a Kalshi market identifier to the
``(settlement_source, settlement_station)`` tuple downstream code uses
to pull the right settlement row from the CLI catalog.
"""

from __future__ import annotations

from dataclasses import dataclass
from datetime import date as _date

from tradewinds.markets.catalog.kalshi_stations import KALSHI_SETTLEMENT_STATIONS


@dataclass(frozen=True)
class NHighResolution:
    """The (source, station) tuple a Kalshi NHIGH contract resolves to."""

    settlement_source: str
    settlement_station: str
    city_ticker: str
    contract_date: _date


def resolve(contract_id: str, settlement_date: _date) -> NHighResolution:
    """Resolve a Kalshi NHIGH contract to its settlement source + station.

    The contract_id is the Kalshi market identifier — for v0.1.0 the
    expected format is ``KHIGH<CITY>`` (e.g. ``KHIGHNY``, ``KHIGHCHI``).
    The city ticker is the suffix; we look it up in the whitelist.

    Args:
        contract_id: Kalshi market identifier. Case-insensitive.
        settlement_date: The local calendar date the market settles for.

    Returns:
        :class:`NHighResolution` with settlement_source = ``"cli.archive"``
        and settlement_station from the whitelist.

    Raises:
        ValueError: ``contract_id`` doesn't follow the ``KHIGH<CITY>`` format
            or the city ticker is not in :data:`KALSHI_SETTLEMENT_STATIONS`.
    """
    # codex iter-8 HIGH fix: validate settlement_date type up front.
    # Accepting arbitrary values lets invalid input escape the resolver
    # boundary and fail later inside the settlement lookup with a
    # confusing message.
    if not isinstance(settlement_date, _date):
        raise TypeError(
            "settlement_date must be a datetime.date instance "
            f"(got {type(settlement_date).__name__}={settlement_date!r})"
        )
    if not isinstance(contract_id, str):
        raise TypeError(
            "contract_id must be a string " f"(got {type(contract_id).__name__}={contract_id!r})"
        )

    cid = contract_id.upper()
    if not cid.startswith("KHIGH") or len(cid) <= 5:
        raise ValueError(f"NHIGH contract_id must follow 'KHIGH<CITY>' format; got {contract_id!r}")
    city_ticker = cid[5:]
    citation = KALSHI_SETTLEMENT_STATIONS.get(city_ticker)
    if citation is None:
        raise ValueError(
            f"Unknown Kalshi city ticker {city_ticker!r}; "
            f"known: {sorted(KALSHI_SETTLEMENT_STATIONS)}"
        )
    return NHighResolution(
        settlement_source="cli.archive",
        settlement_station=citation.station,
        city_ticker=city_ticker,
        contract_date=settlement_date,
    )


__all__ = ["NHighResolution", "resolve"]

File: packages-ts/markets/src/resolvers/kalshi-nhigh.ts
// Kalshi NHIGH (daily HIGH temperature) contract resolver.
//
// NHIGH markets resolve against the NWS CLI `max_temp_f` value for a
// specific station on a specific date. `kalshiNhighResolve` is the
// deterministic mapping from a Kalshi market identifier to the
// (settlement_source, settlement_station) tuple downstream code uses
// to pull the right settlement row from the CLI catalog.
//
// Ported byte-faithful from packages/markets/src/mostlyright/markets/catalog/kalshi_nhigh.py.

import { KALSHI_SETTLEMENT_STATIONS } from "../data/generated/kalshi-stations.js";

export interface NHighResolution {
  readonly settlementSource: "cli.archive";
  readonly settlementStation: string; // 4-letter ICAO
  readonly cityTicker: string;
  readonly contractDate: string; // YYYY-MM-DD
}

/**
 * Custom error type for contract-id parsing / validation failures.
 *
 * Mirrors the Python `ValueError`/`TypeError` distinction: in TS we use a
 * named subclass so callers can `instanceof`-check rather than parse
 * error messages.
 */
export class ContractIdError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "ContractIdError";
  }
}

/**
 * Coerce a Date or YYYY-MM-DD string into an ISO date-only string.
 *
 * Rejects Date instances whose UTC hours/minutes/seconds/ms are non-zero —
 * those would silently corrupt downstream settlement-date matching
 * (date-equality is strict).
 *
 * Mirrors the Python `isinstance(settlement_date, datetime)` guard in
 * `kalshi_nhigh.resolve`.
 */
function coerceContractDate(value: Date | string): string {
  if (typeof value === "string") {
    // YYYY-MM-DD strict format check.
    if (!/^\d{4}-\d{2}-\d{2}$/.test(value)) {
      throw new ContractIdError(
        `settlementDate string must be YYYY-MM-DD; got ${JSON.stringify(value)}`,
      );
    }
    // Validate it parses to a real calendar date.
    const parsed = new Date(`${value}T00:00:00Z`);
    if (Number.isNaN(parsed.getTime())) {
      throw new ContractIdError(
        `settlementDate string is not a valid calendar date; got ${JSON.stringify(value)}`,
      );
    }
    // Round-trip check — guards against e.g. "2025-02-30" silently
    // rolling forward.
    const iso = parsed.toISOString().slice(0, 10);
    if (iso !== value) {
      throw new ContractIdError(
        `settlementDate string is not a valid calendar date; got ${JSON.stringify(value)}`,
      );
    }
    return value;
  }
  if (value instanceof Date) {
    if (Number.isNaN(value.getTime())) {
      throw new ContractIdError("settlementDate is an invalid Date instance");
    }
    // Reject any Date with a non-zero UTC time component — mirrors
    // Python's `isinstance(settlement_date, datetime)` guard. A Date
    // carries a time component which would break downstream
    // date-equality matching.
    if (
      value.getUTCHours() !== 0 ||
      value.getUTCMinutes() !== 0 ||
      value.getUTCSeconds() !== 0 ||
      value.getUTCMilliseconds() !== 0
    ) {
      throw new ContractIdError(
        `settlementDate must be a UTC date-only Date (H/M/S/ms = 0); got ${value.toISOString()}. Use new Date('YYYY-MM-DDT00:00:00Z') or pass a 'YYYY-MM-DD' string.`,
      );
    }
    return value.toISOString().slice(0, 10);
  }
  throw new ContractIdError("settlementDate must be a Date instance or YYYY-MM-DD string");
}

/**
 * Resolve a Kalshi NHIGH contract to its settlement source + station.
 *
 * The contract id format is `KHIGH<CITY>` (case-insensitive), where
 * `<CITY>` is a city ticker present in
 * {@link KALSHI_SETTLEMENT_STATIONS}.
 *
 * @param contractId Kalshi market identifier. Case-insensitive.
 * @param settlementDate Calendar date the market settles for. Either a
 *   UTC date-only `Date` (H/M/S/ms == 0) or a `YYYY-MM-DD` string.
 * @returns A frozen {@link NHighResolution}.
 * @throws {ContractIdError} The contract id doesn't follow
 *   `KHIGH<CITY>`, the city is unknown, or the settlement date is
 *   invalid.
 */
export function kalshiNhighResolve(
  contractId: string,
  settlementDate: Date | string,
): NHighResolution {
  if (typeof contractId !== "string") {
    throw new ContractIdError(`contractId must be a string; got ${typeof contractId}`);
  }

  const contractDate = coerceContractDate(settlementDate);

  const cid = contractId.toUpperCase();
  if (!cid.startsWith("KHIGH") || cid.length <= 5) {
    throw new ContractIdError(
      `NHIGH contractId must follow 'KHIGH<CITY>' format; got ${JSON.stringify(contractId)}`,
    );
  }
  const cityTicker = cid.slice(5);
  const station = KALSHI_SETTLEMENT_STATIONS[cityTicker];
  if (station === undefined) {
    const known = Object.keys(KALSHI_SETTLEMENT_STATIONS).sort();
    throw new ContractIdError(
      `unknown city ${JSON.stringify(cityTicker)} (known: ${known.join(", ")})`,
    );
  }

  return Object.freeze({
    settlementSource: "cli.archive" as const,
    settlementStation: station.station,
    cityTicker,
    contractDate,
  });
}
